### ISSUE TITLE
transition to nloptr BOBYQA?

### ISSUE THREAD

**Issue by bbolker** [lme4/lme4#484]
Quick thoughts/brain dump:

- we should check carefully whether the difference between minqa and nloptr implementations of BOBYQA is partly or largely due to differences in the default tolerances; I remember comments from Gabor Grothendieck along these lines ...
- need to track down reasons for @mmaechler's difficulties with the `Pastes` example. Even if this is a fake example, I don't think it's so pathological that we can completely discount it ... FWIW I just updated to `nloptr` 1.2.1, still not reproducing the problems
- when I run the tests I get lots of gradient warnings. These can obviously be cleaned up by cheating (i.e. suppressing warnings/lowering tolerances). This is related to the whole toast-scraping issue, but we might want to take this as an opportunity to figure out some more of what's going on ... (i.e. in this case I guess we get gradient warnings with `nloptr` BOBYQA but not with `minqa` BOBYQA?)
- FWIW it looks like recent versions of `nloptr` do *not* require network connection to install - they come bundled with the relevant source codes from NLOpt ...

**Comment by mmaechler** [comment]
In the cases where I had changed tolerances (or target values) for the tests,  I've looked at the objective function values (i.e.  -2 logLik or REML)  and typically  (about 6 to 1 ?)  the `nloptr` based result both converged more quickly *and* had reached a better value of the objective function.

OTOH, I hadn't found the time to investiagate my `Pastes`+"nloptr"  problems which you did not see.

**Comment by bachlaw** [comment]
I may be jumping in late but I cannot think of an example where the nloptr optimizer (at least for glmer) has not been substantially faster while providing what appear to be at least equivalent results. I have made its usage standard in my workflow for quite some time now. There is just no reason i can see to wait for the standard optimizer, particularly with large data sets. Other applications may vary of course.

Jonathan

Sent from my iPhone

On Oct 6, 2018, at 3:06 AM, Martin Mächler <notifications@github.com<mailto:notifications@github.com>> wrote:


In the cases where I had changed tolerances (or target values) for the tests, I've looked at the objective function values (i.e. -2 logLik or REML) and typically (about 6 to 1 ?) the nloptr based result both converged more quickly and had reached a better value of the objective function.

OTOH, I hadn't found the time to investiagate my Pastes+"nloptr" problems which you did not see.

**Comment by dmbates** [comment]
I think that @bachlaw has a telling point.  In my recollection there have been many instances where the nloptr optimizer is faster than the current default and has converged to a better value.  I don't know of instances other than the experience of @mmaechler with the `Pastes` example where the current default has been superior.

As far as the gradient warnings go, I have seen many, many false positives from those checks.  I don't think I have seen "true positives".  Some of the checks are clearly wrong - when a constrained optimizer converges to a boundary value you do **not** expect the gradient to be approximately zero and it is not a mistake if the Hessian is not positive definite.  I think the onus should be on showing that they are useful if they are to be retained.

**Comment by bbolker** [comment]
We've switched now, in the master branch.  We'll see what the fallout is in reverse dependencies etc. (a heads-up message to `r-sig-mixed-models` would probably be a good idea; also a `NEWS` entry).  FWIW the gradient warnings are no longer assessed when we detect that we've got a singular fit.  And I **have** seen true positives, usually as indications that someone's doing something really silly.  I suspect that @dmbates doesn't try to fit as many truly crazy models as the general public does ...
